### DISCUSSION TITLE
Cache headers for 404's in ISR/dynamic pages

### DISCUSSION THREAD

**Discussion by jonjohansen** [vercel/next.js#35218] [Help] [answered]
Hello 👋 

I'm working on an application which has various dynamic routes that are built with ISR, with a blocking fallback. Essentially this means that pages could, and should be published from a CMS while the application is running. This works super nice - for the most part. An issue which I've found seems to be the cache headers that are given when the page hasn't been published yet and are visited, then published afterwards. 

Upon not finding any results (`404`) the response has a  cache header `Cache-Control: s-maxage=31536000, stale-while-revalidate` - Which would make a lot of sense if the pages were static. 
However if a user navigates to a specific path before the page is published, it is cached as a `404` for a very long time, while in reality it could be published any minute!  I expected the cache behaviour to be similar to the one of the hits - Cached for the same amount as the revalidation interval.

Similarly if using a `next/link` with prefetch, the 404 on a missing `.json` file corresponding to a missing page in `_next/data/<buildId>/` is also cached. This would however be resolved with a rebuild - as the `buildId` part changes and effectively invalidates the cache.

This issue is even harder to work around when using an external CDN which cache these results based on the cache headers provided from the original request. For each published page i have to invalidate the CDN cache manually, in case it has been cached as a `404` previously - and thus is cached for a whole year.

I'm wondering if there is some way of overriding the cache headers for dynamic pages (which should be the same as the other revaldation-intervals), and potentially within folders for the prefetch-data, to avoid a year-long `404` being cached. 

Perhaps there's something i've missed in the documentation 😬

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I had initially answered thinking that you had `getServerSideProps`, but I now realise that you are probably using `getStaticPaths` and `getStaticProps` working together.

```js
export const Post = () => {
  return <div>Post</div>;
};

export const getStaticPaths = () => {
  return {
    fallback: true,
    paths: [{ params: { slug: "a" } }],
  };
};

/** @type {import('next').GetStaticProps} */
export const getStaticProps = (ctx) => {
  console.log("Post");

  if (ctx.params.slug === "a") {
    return { props: {} };
  }

  if (Math.random() > 0.5) {
    return { props: {}, revalidate: 10 };
  }

  return {
    notFound: true,
    // revalidate: 10,
  };
};

export default Post;
``` 

In the above, when I don't return `revalidate: 10` on not found, then the 404 is cached for `Cache-Control: s-maxage=31536000, stale-while-revalidate`, though `stale-while-revalidate` means that it should be fetching a new version on the background, unfortunately it does that after `31536000`.

However, if I uncomment the `revalidate: 10`, and return it with the not found, then the cache is only there for 10 seconds, and `stale-while-revalidate` actually fetches a new version after 10 seconds. On the snippet above, a new `console.log("Post")` is executed after the 10 seconds. 

Does that fix it for you?

**Reply by jonjohansen** [reply]
Woah! Thank you so much. That solved it! 👏 👏  It effectively set the cache header for `404` in both the `.json` and the document to the set interval.

I had no idea that you could use the `revalidate` in combination with `notFound`.  
Maybe this is something that could be more explicitly stated in the documentation of [Cache-control](https://github.com/vercel/next.js/discussions/35218) headers, or perhaps an example in the documentation regarding ISR?

Thanks so much for a quick reply! 😸

**Reply by icyJoseph** [reply]
Yeah I think this should be more clear in the docs.

**Reply by ng-hai** [reply]
OMG you save my day @icyJoseph
